### ISSUE TITLE
Gradient not working with tf.math.angle

### ISSUE THREAD

**Issue by NEGU93** [tensorflow/tensorflow#50854]
I tried to create my own loss function that uses `tf.math.angle` in it. However, when I use this custom loss I get `None` gradients.
I created a minimum working example (link [here](https://colab.research.google.com/drive/1SkomH1STPtHX6GfB8RBDRuu-pK5VkTyk?usp=sharing)).

The expected behavior for this example would be the gradient to be zero as the loss is zero so there is no minimizing it.
Another note is that the loss would be zero for every real number which is what we are going to have using Tensorflow.

My motivation is because I work with complex numbers as you can see [here](https://github.com/NEGU93/cvnn). So this loss function will make sense on the complex domain as not every tensor will have the same angle.
This issue is related to [this](https://github.com/NEGU93/cvnn/issues/12).

**Comment by UsharaniPagadala** [comment]
@Saduf2019 

I  was able to replicate the issue reported here. Please find the [gist](https://colab.research.google.com/gist/UsharaniPagadala/7925e19240055a24668f4a29de195583/copy-of-loss_with_angle.ipynb).Thanks

**Comment by MaanasArora** [comment]
Hi @Saduf2019, @NEGU93:

I tried out the code in @UsharaniPagadala's gist--it seems like we're computing `tf.math.angle` of `tf.float32` instances, which requires that an `array_ops.where` operation be done (and it's not useful because the results are always zero anyway). This disconnects the gradient, resulting in None `grads`. 

If you replace this line:
`angle_diff = tf.math.angle(y_true) - tf.math.angle(y_pred)`
with this (which we don't want, but for illustration):
`angle_diff = tf.math.angle(tf.complex(y_true, 0.)) - tf.math.angle(tf.complex(y_true, 0))`
we get tensor gradients.

For complex number applications, I assume we have to explicitly declare the `dtype` of the input so that we don't run into this problem? I'm a new contributor, so let me know if I have this right. Thanks

**Comment by NEGU93** [comment]
@MaanasArora  I tried the change you suggested in the gist but still have the same error... Can you provide an example?

**Comment by MaanasArora** [comment]
@NEGU93 With this code?

```python
def mean_squared_angle_err(y_true, y_pred):
    angle_diff = tf.math.angle(y_true) - tf.math.angle(y_pred)
    angle_diff = ((angle_diff + np.pi) % (2*np.pi)) - np.pi # [-pi:pi]
    return tf.math.reduce_mean(angle_diff**2)
```

Sorry, this 'worked' for me but will not return valid losses, I just provided it for illustration.

**Comment by NEGU93** [comment]
I am not sure I understand, you said that if I replaced with the line `angle_diff = tf.math.angle(tf.complex(y_true, 0.)) - tf.math.angle(tf.complex(y_true, 0))` we should have gradients but I didn't. Is there something else that should be changed for it to work? Why is it not working for me?

Regards,

**Comment by MaanasArora** [comment]
Sorry, I edited a few typos in the code above. Could you try again?

**Comment by NEGU93** [comment]
If it's the point after the zero I corrected that when I tried it not to have the error about the int32. Still, I keep having the None gradients.

**Comment by NEGU93** [comment]
Here is the [gist](https://colab.research.google.com/gist/NEGU93/af072c1b50530ad4e868348d09b0f5ea/copy-of-loss_with_angle.ipynb) with the changes. You can see the gradient is still None.

**Comment by MaanasArora** [comment]
Did you replace the second `y_true` with `y_pred`?

**Comment by MaanasArora** [comment]
I just meant that the gradients become non-None python values--if you want not NAN numpy values you can replace the 0s by 1s. The angle for a complex number with a zero imaginary component is NAN. This is not a real solution.

**Comment by NEGU93** [comment]
Yes, I changed it to 1 and worked!

The angle with a zero imaginary part is NAN? That is a weird behavior. I wonder the reason for that but I guess that is for another issue.

**Comment by MaanasArora** [comment]
As for the code in the complex number gist, the model outputs are `float32` instances which explains why their output returns nans. I assume the CVNN library might be using floats to construct complex numbers, in which case you will need to deconstruct the complex number and call `tf.complex`. That's why I suspect the issue lies outside of tensorflow. This illustration will not return valid losses.

**Comment by NEGU93** [comment]
Well, that is a definition, however, atan(b/a) = atan(0/a) = 0 and not Nan, the problem would be a zero real part (wich is also bad).
Also, using that definition will have the problem of -a-ib (third quadrant) will have the angle of a+ib (first quadrant) as -a/-b = a/b right?.

I would expect a complex number with a zero imaginary part (a+i0) would be on the real axis, therefore, angle = 0.

Moreover, I tried just computing:
`tf.math.angle(tf.complex(tf.ones((1,1)), 0.))` and it does return `0` and not Nan. 

The problem was indeed solved as you said but I still fail to understand it.

**Comment by NEGU93** [comment]
Ups, stupid me, I thought inp was ones and not zeros, of course a 0+i0 would be a problem. I got it now! Thank you very much! 

I will close the issue then.

**Comment by NEGU93** [comment]
Sorry, opened it again, if I do `tf.math.angle(tf.zeros((1, 1), dtype=np.complex64))` it gives me 0 and not Nan, so apparently, TF has chosen to make 0+i0 angle to be zero (sounds ok and actually a good choice), and therefore the issue it is still not explained for me.

**Comment by MaanasArora** [comment]
Sorry, since I am not a member of this repository, my understanding might not be definitive. 

The `None` gradients resulting from a disconnected graph, caused by `float` inputs to `tf.math.angle`, would be an issue because the loss will not compute. What I can gather is that `nan` numpy values for gradients are acceptable because those gradients are _differentials_ of zero values, and this shouldn't cause an issue when actual complex numbers are inserted.

**Comment by NEGU93** [comment]
No need to go to the repository, that is my job once this is cleared. My question is with the original gist.
We have the following remarks:

1. using `angle_diff = tf.math.angle(tf.complex(y_true, 0.)) - tf.math.angle(tf.complex(y_pred, 0.))` gives nan gradients.
2. Using `angle_diff = tf.math.angle(tf.complex(y_true, 1.)) - tf.math.angle(tf.complex(y_pred, 1.))` appears to work.

The reason behind this is still unknown to me. even if `y_true` and/or `y_pred` is zero, tensorflow appears to return a `0` angle and not `Nan` as one can see while calling `tf.math.angle(tf.zeros((1, 1), dtype=np.complex64))`. So why would I have nan gradients? 
Why does `0+i0` break the gradients if tensorflow appears to interpret the angle of `0+i0` to equal `0`?

**Comment by NEGU93** [comment]
Shall I close this and open another issue asking why the gradient of `0+i0` is `nan` or we continue the chat in this one? I still don't get why

> `nan` numpy values for gradients are acceptable when those gradients are _differentials_ of zero values, and this shouldn't cause an issue when actual complex numbers are inserted.

**Comment by Saduf2019** [comment]
@NEGU93 
Tf.math angle is configured such that it will return 0 instead of nan, the operation must support gradients which is not possible if the value becomes nan or inf.

**Comment by NEGU93** [comment]
Sorry for this. But I still don't get it.

1. Gradients do not work if operation returns nan of inf
2. This case is coded to return 0, so no nan nor inf.

Then what is the problem?

**Comment by Saduf2019** [comment]
@NEGU93 
That is what it is supposed to do tf.math.angle must support gradients and since gradients are not defined for inf and nan. Kindly move this to closed status and open an issue on tf discussion group for any further queries as this is not a bug or feature request and there is a larger community there to respond. Thanks!

**Comment by NEGU93** [comment]
Ok. I will do that. Thank you!
